presto-memory-context/src/main/java/com/facebook/presto/memory/context/AggregatedMemoryContext.java

```diff
@@ -0,0 +1,151 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.memory.context;
+
+import com.google.common.util.concurrent.ListenableFuture;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.OptionalLong;
+
+import static com.facebook.presto.memory.context.MemoryReservationHandler.NOT_BLOCKED;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class AggregatedMemoryContext
+{
+    @Nullable
+    final AggregatedMemoryContext parentMemoryContext;
+    @Nullable
+    private final MemoryReservationHandler reservationHandler;
+    @GuardedBy("this")
+    long usedBytes;
+    @GuardedBy("this")
+    private boolean closed;
+    private final OptionalLong guaranteedMemory;
+
+    public AggregatedMemoryContext()
+    {
+        this(null, OptionalLong.empty());
+    }
+
+    public AggregatedMemoryContext(MemoryReservationHandler reservationHandler, OptionalLong guaranteedMemoryInBytes)
+    {
+        this(null, reservationHandler, guaranteedMemoryInBytes);
+    }
+
+    private AggregatedMemoryContext(AggregatedMemoryContext parentMemoryContext, MemoryReservationHandler reservationHandler, OptionalLong guaranteedMemory)
+    {
+        checkArgument(reservationHandler == null || parentMemoryContext == null, "at least one of reservationHandler and parentMemoryContext must be null");
+        this.parentMemoryContext = parentMemoryContext;
+        this.reservationHandler = reservationHandler;
+        this.guaranteedMemory = requireNonNull(guaranteedMemory, "guaranteedMemory is null");
+    }
+
+    public AggregatedMemoryContext newAggregatedMemoryContext()
+    {
+        return new AggregatedMemoryContext(this, null, OptionalLong.empty());
+    }
+
+    public LocalMemoryContext newLocalMemoryContext()
+    {
+        return new SimpleLocalMemoryContext(this);
+    }
+
+    public synchronized long getBytes()
+    {
+        return usedBytes;
+    }
+
+    synchronized ListenableFuture<?> updateBytes(long bytes)
+    {
+        ListenableFuture<?> future = NOT_BLOCKED;
+        // delegate to parent if exists
+        if (parentMemoryContext != null) {
+            // update the parent before updating usedBytes as it may throw a runtime exception (e.g., ExceededMemoryLimitException)
+            future = parentMemoryContext.updateBytes(bytes);
+            usedBytes += bytes;
+            return future;
+        }
+
+        // At this point this is root
+        if (reservationHandler != null) {
+            future = reservationHandler.reserveMemory(bytes);
+        }
+
+        usedBytes += bytes;
```

Comment:
Can this overflow (e.g. when `bytes` is a bogus value miscalculated by the caller)?
Can `usedBytes` become negative (when `abs(bytes) > abs(usedBytes)`)?